#!/usr/bin/env python3
"""
AIæœåŠ¡æ€§èƒ½åŸºçº¿æµ‹è¯•å·¥å…·
"""

import time
import statistics
import grpc
import sys
import os

# æ·»åŠ å½“å‰ç›®å½•åˆ°Pythonè·¯å¾„
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

import intelligence_pb2
import intelligence_pb2_grpc

def measure_ai_performance():
    """æµ‹é‡AIæœåŠ¡å½“å‰æ€§èƒ½åŸºçº¿"""
    print("ğŸ” AIæœåŠ¡æ€§èƒ½åŸºçº¿æµ‹é‡")
    print("=" * 50)
    
    # è¿æ¥åˆ°AIæœåŠ¡
    channel = grpc.insecure_channel('localhost:50051')
    stub = intelligence_pb2_grpc.IntelligenceServiceStub(channel)
    
    test_cases = [
        "æ›´æ–°æˆ‘çš„ç”µè¯å·ç ä¸º13800138000",
        "è°æ˜¯æˆ‘çš„ç»ç†ï¼Ÿ",
        "æˆ‘æƒ³æŸ¥çœ‹æˆ‘çš„ä¸ªäººä¿¡æ¯",
        "å¸®æˆ‘ç”³è¯·å¹´å‡",
        "æŸ¥è¯¢å…¬å¸ç»„ç»‡æ¶æ„"
    ]
    
    response_times = []
    
    print(f"æµ‹è¯•ç”¨ä¾‹æ•°é‡: {len(test_cases)}")
    print("å¼€å§‹æ€§èƒ½æµ‹è¯•...\n")
    
    for i, text in enumerate(test_cases, 1):
        print(f"æµ‹è¯• {i}/{len(test_cases)}: {text[:20]}...")
        
        start_time = time.time()
        try:
            request = intelligence_pb2.InterpretRequest()
            request.user_text = text
            request.session_id = f"perf-test-{i}"
            
            response = stub.InterpretText(request, timeout=30)
            end_time = time.time()
            
            response_time = (end_time - start_time) * 1000  # è½¬æ¢ä¸ºæ¯«ç§’
            response_times.append(response_time)
            
            print(f"  âœ… å“åº”æ—¶é—´: {response_time:.0f}ms")
            print(f"  ğŸ“‹ æ„å›¾: {response.intent}")
            
        except Exception as e:
            print(f"  âŒ æµ‹è¯•å¤±è´¥: {e}")
    
    # ç»Ÿè®¡åˆ†æ
    if response_times:
        print("\n" + "=" * 50)
        print("ğŸ“Š æ€§èƒ½ç»Ÿè®¡ç»“æœ:")
        print(f"  å¹³å‡å“åº”æ—¶é—´: {statistics.mean(response_times):.0f}ms")
        print(f"  æœ€çŸ­å“åº”æ—¶é—´: {min(response_times):.0f}ms")
        print(f"  æœ€é•¿å“åº”æ—¶é—´: {max(response_times):.0f}ms")
        print(f"  å“åº”æ—¶é—´ä¸­ä½æ•°: {statistics.median(response_times):.0f}ms")
        if len(response_times) > 1:
            print(f"  æ ‡å‡†å·®: {statistics.stdev(response_times):.0f}ms")
        
        # æ€§èƒ½è¯„ä¼°
        avg_time = statistics.mean(response_times)
        print(f"\nğŸ¯ æ€§èƒ½ç›®æ ‡åˆ†æ:")
        print(f"  å½“å‰å¹³å‡å“åº”æ—¶é—´: {avg_time:.0f}ms")
        print(f"  ç›®æ ‡å“åº”æ—¶é—´: <2000ms")
        if avg_time > 2000:
            improvement_needed = ((avg_time - 2000) / avg_time) * 100
            print(f"  éœ€è¦æ”¹è¿›: {improvement_needed:.1f}%")
            print(f"  çŠ¶æ€: âŒ éœ€è¦ä¼˜åŒ–")
        else:
            print(f"  çŠ¶æ€: âœ… å·²è¾¾æ ‡")
    
    channel.close()
    return response_times

if __name__ == "__main__":
    try:
        response_times = measure_ai_performance()
    except Exception as e:
        print(f"âŒ æ€§èƒ½æµ‹è¯•å¤±è´¥: {e}")
        sys.exit(1)